iib/workers/tasks/build_recursive_related_bundles.py

```diff
@@ -0,0 +1,160 @@
+            temp_current_level_related_bundles = copy.deepcopy(current_level_related_bundles)
+            current_level_related_bundles = []
+            for bundle in temp_current_level_related_bundles:
+                children_related_bundles = process_parent_bundle_image(
+                    bundle, request_id, organization
+                )
+                current_level_related_bundles.extend(children_related_bundles)
+
+                total_related_bundles += len(children_related_bundles)
+                if total_related_bundles >= conf['iib_max_recursive_related_bundles']:
+                    raise IIBError('Max number of related bundles exceeded. Potential DDOS attack!')
+
+            recursive_related_bundles.extend(current_level_related_bundles)
+            if not current_level_related_bundles:
+                traversal_completed = True
+
+    payload = {
+        'state': 'in_progress',
+        'state_reason': 'Writing recursive related bundles to a file',
+    }
+    update_request(request_id, payload, exc_msg='Failed setting the bundle image on the request')
+    # Reverse the list while writing because we did a top to bottom level traversal of a tree.
+    # The return value should be a bottom to top level traversal.
+    write_related_bundles_file(
+        recursive_related_bundles[::-1],
+        request_id,
+        conf['iib_request_recursive_related_bundles_dir'],
+        'recursive_related_bundles',
+    )
+
+    payload = {
+        'state': 'complete',
+        'state_reason': 'The request completed successfully',
+    }
+    update_request(request_id, payload, exc_msg='Failed setting the bundle image on the request')
+
+
+def process_parent_bundle_image(
+    bundle_image_resolved: str, request_id: int, organization: Optional[str] = None
+) -> List[str]:
+    """
+    Apply required customization and get children bundles (aka related bundles) for a bundle image.
+
+    :param str bundle_image_resolved: the pull specification of the bundle image to whose children
+        bundles are to be found.
+    :param int request_id: the ID of the IIB build request
+    :param str organization: the name of the organization the to apply customizations on.
+    :rtype: list
+    :return: the list of all children bundles for a parent bundle image
+    :raises IIBError: if fails to process the parent bundle image.
+    """
+    # Pull the bundle_image to ensure steps later on don't fail due to registry timeouts
+    podman_pull(bundle_image_resolved)
+
+    with tempfile.TemporaryDirectory(prefix='iib-') as temp_dir:
+        manifests_path = os.path.join(temp_dir, 'manifests')
+        _copy_files_from_image(bundle_image_resolved, '/manifests', manifests_path)
+        metadata_path = os.path.join(temp_dir, 'metadata')
+        _copy_files_from_image(bundle_image_resolved, '/metadata', metadata_path)
+        if organization:
+            _adjust_operator_bundle(manifests_path, metadata_path, request_id, organization)
```

Comment:
I assume that you want to do this so that you can modify any pullspecs to ensure that they are actually pullable so that you can then `get_related_bundle_images` later. 

This will process all operations requested for the organization. If an organization has multiple `registry_replacements` customizations specified, however, intermediate actions to make images pullable might be overruled by later operations to change the registry to a final (and not yet valid) pullspec. 

Assuming that the organizations are properly configured, they can specify the `related_bundles` type at a location where the images _can_ be pulled. We should have that process cache off the related bundles so that we can then read from that cache when we call it again lower.

Reply:
Out of band, you commented that we don't need the pullspecs to be resolvable, but we do. As part of the later function call `get_related_bundle_images(bundle_metadata)`, we need to be able to get the image labels to determine whether the related image is a bundle or not:

```python
    for related_pullspec_obj in bundle_metadata['found_pullspecs']:
        related_pullspec = related_pullspec_obj.to_str()
        if yaml.load(
            get_image_label(related_pullspec, 'com.redhat.delivery.operator.bundle') or 'false'
        ):
            related_bundle_images.append(related_pullspec)
```

Reply:
As discussed on the call, I have made the changes to terminate adjusting the opertor bundle when `related_bundles` customization is reached.